File: solver_sdk/ios/Sources/PolyPuzzleSolverSDK/SolverService.swift
import Foundation
import JavaScriptCore

/// The single entry point SwiftUI (or any Application Layer code) calls to
/// run the Solver Engine. Wraps a JSContext that evaluates
/// PolyPuzzleSolverBridge.bundle.js -- the esbuild output of
/// solver_sdk/bridge/PolyPuzzleSolverBridge.ts, which itself imports
/// FiveByFiveEdgeSolverEngine directly from src/customCube (frozen Baseline
/// V1.0, never modified or reimplemented). This class contains NO solving
/// logic of its own -- it only marshals JSON in, calls the one bridge
/// function, and marshals JSON back out.
///
/// Why JavaScriptCore + a bundled JS file, not a Swift port: rewriting the
/// Solver Engine in Swift would mean re-validating everything the
/// Continuous Validation Framework / Long-term Reliability Validation
/// Sprints already measured against the TypeScript implementation, and
/// would put the iOS app permanently out of sync with any future
/// Maintenance Sprint fix (docs/SOLVER_MAINTENANCE_POLICY.md) unless that
/// fix were manually ported every time. JavaScriptCore ships on every iOS
/// device at zero extra dependency cost and lets the SDK run the exact same
/// bits the web app and every prior Sprint's benchmark ran.
public final class SolverService {
  private let context: JSContext
  private let encoder: JSONEncoder
  private let decoder: JSONDecoder

  /// Throws `.bridgeUnavailable` if the bundled resource is missing or
  /// fails to evaluate -- e.g. a broken app bundle. `bundle` defaults to
  /// `Bundle.module`, the resource bundle Swift Package Manager generates
  /// for this target's `Resources/` folder (declared in Package.swift).
  public init(bundle: Bundle = .module) throws {
    guard let context = JSContext() else {
      throw SolverError.bridgeUnavailable(message: "JSContext() returned nil")
    }
    guard let scriptURL = bundle.url(forResource: "PolyPuzzleSolverBridge.bundle", withExtension: "js") else {
      throw SolverError.bridgeUnavailable(message: "PolyPuzzleSolverBridge.bundle.js not found in bundle resources")
    }
    let script: String
    do {
      script = try String(contentsOf: scriptURL, encoding: .utf8)
    } catch {
      throw SolverError.bridgeUnavailable(message: "failed to read bundle resource: \(error.localizedDescription)")
    }

    var caughtException: JSValue?
    context.exceptionHandler = { _, exception in caughtException = exception }
    context.evaluateScript(script)
    if let caughtException {
      throw SolverError.bridgeUnavailable(message: "bundle evaluation threw: \(caughtException.toString() ?? "(unknown)")")
    }
    guard context.objectForKeyedSubscript("PolyPuzzleSolverBridge")?.isUndefined == false else {
      throw SolverError.bridgeUnavailable(message: "PolyPuzzleSolverBridge global not defined after evaluating bundle")
    }

    self.context = context
    self.encoder = JSONEncoder()
    self.decoder = JSONDecoder()
  }

  /// Pre-builds the Solver Engine's wing/flip/case libraries (mirrors the
  /// web app's own warmup() call site in App.tsx) -- call once at app
  /// launch or when the Solve screen first appears, well before the user
  /// presses Solve, so the ~1.3s cold-build cost documented in
  /// fiveByFiveEdgeSolverEngine.ts doesn't eat into the first real solve's
  /// own 1000ms plan budget.
  public func warmup() {
    context.objectForKeyedSubscript("PolyPuzzleSolverBridge")?.invokeMethod("warmup", withArguments: [])
  }

  /// Runs one solve() call against the frozen Solver Engine.
  ///
  /// Timing note (disclosed, not a defect -- see
  /// docs/IOS_PRODUCT_ARCHITECTURE.md §Bridge Verification): solve() is a
  /// wall-clock time-budgeted algorithm (PLAN_TIME_BUDGET_MS=1000ms,
  /// hardcoded, see docs/SOLVER_BASELINE_V1.md). Its OUTPUT for a given
  /// input can therefore vary between two otherwise-identical calls if the
  /// host CPU's real execution speed differs -- this was already
  /// established and disclosed for the web app itself in
  /// docs/SOLVER_LONG_TERM_RELIABILITY_VALIDATION.md's Determinism
  /// Analysis. This SDK does not change or hide that property: on-device
  /// JavaScriptCore performance (not measured in this Sprint -- no
  /// physical device or simulator was available in this Sprint's Linux
  /// container) determines exactly how much of a hard scramble's plan
  /// completes within the same 1000ms window, same as it does for any
  /// other host.
  public func solve(_ request: SolverRequest) throws -> SolveResult {
    guard let bridge = context.objectForKeyedSubscript("PolyPuzzleSolverBridge"),
      !bridge.isUndefined
    else {
      throw SolverError.bridgeUnavailable(message: "PolyPuzzleSolverBridge global missing at call time")
    }

    let requestData: Data
    do {
      requestData = try encoder.encode(request)
    } catch {
      throw SolverError.malformedResponse(message: "failed to encode SolverRequest: \(error.localizedDescription)")
    }
    guard let requestJson = String(data: requestData, encoding: .utf8) else {
      throw SolverError.malformedResponse(message: "SolverRequest JSON was not valid UTF-8")
    }

    var caughtException: JSValue?
    context.exceptionHandler = { _, exception in caughtException = exception }
    guard let responseValue = bridge.invokeMethod("solve", withArguments: [requestJson]),
      let responseJson = responseValue.toString()
    else {
      let detail = caughtException?.toString() ?? "(no exception captured, no return value)"
      throw SolverError.bridgeUnavailable(message: "PolyPuzzleSolverBridge.solve() call failed: \(detail)")
    }

    guard let responseData = responseJson.data(using: .utf8) else {
      throw SolverError.malformedResponse(message: "bridge response was not valid UTF-8")
    }

    // The bridge's own contract (PolyPuzzleSolverBridge.ts's solveBridge())
    // is `{ok:true,...} | {ok:false,errorCode,errorMessage}` -- decode the
    // discriminator first via the lightweight error-shaped struct, then
    // decode the full SolveResult only on the success path, rather than
    // guessing from which decode succeeds.
    guard let okFlag = try? decoder.decode(OkFlagOnly.self, from: responseData) else {
      throw SolverError.malformedResponse(message: "response JSON had no decodable `ok` field: \(responseJson.prefix(200))")
    }
    if !okFlag.ok {
      guard let errorPayload = try? decoder.decode(SolverBridgeErrorPayload.self, from: responseData) else {
        throw SolverError.malformedResponse(message: "ok:false response did not match SolverBridgeErrorPayload: \(responseJson.prefix(200))")
      }
      throw SolverError.from(errorPayload)
    }
    do {
      return try decoder.decode(SolveResult.self, from: responseData)
    } catch {
      throw SolverError.malformedResponse(message: "ok:true response did not match SolveResult: \(error.localizedDescription)")
    }
  }
}

private struct OkFlagOnly: Codable {
  var ok: Bool
}

File: mobile/lib/solver/solver_service.dart
import 'dart:convert';

import 'package:flutter/services.dart' show rootBundle;
import 'package:flutter_js/flutter_js.dart';

import 'solver_models.dart';

/// The single entry point the app calls to run the Solver Engine. Wraps a
/// `flutter_js` JavascriptRuntime that evaluates
/// assets/js/PolyPuzzleSolverBridge.bundle.js -- the exact same esbuild
/// output solver_sdk/bridge/build.mjs produces from
/// PolyPuzzleSolverBridge.ts, which itself imports
/// FiveByFiveEdgeSolverEngine directly from src/customCube (frozen
/// Baseline V1.0, never modified). This class contains NO solving logic --
/// it only marshals JSON in, calls the bridge's one function, marshals
/// JSON back out.
///
/// `flutter_js` runs JavaScriptCore on iOS/macOS and QuickJS (via FFI) on
/// Android/Linux/Windows -- one plugin abstraction instead of a
/// per-platform native SDK, which is why this Sprint's Flutter pivot
/// doesn't need an iOS-specific Swift SDK layer the way the previous
/// (superseded) native attempt did. Both platforms still run the identical
/// bundled JS, so the "never reimplement the Solver Engine" guarantee from
/// docs/IOS_PRODUCT_ARCHITECTURE.md's original architecture decision holds
/// unchanged.
class SolverService {
  late final JavascriptRuntime _runtime;
  bool _bridgeReady = false;

  /// Loads and evaluates the bundle. Must be called once (e.g. in main())
  /// before solve()/warmup() -- throws SolverException('BRIDGE_UNAVAILABLE', ...)
  /// if the asset is missing or evaluation fails, mirroring
  /// SolverService.init's own failure mode in the superseded Swift SDK.
  Future<void> initialize() async {
    final bundleSource = await rootBundle.loadString('assets/js/PolyPuzzleSolverBridge.bundle.js');
    _runtime = getJavascriptRuntime();
    final result = _runtime.evaluate(bundleSource, sourceUrl: 'PolyPuzzleSolverBridge.bundle.js');
    if (result.isError) {
      throw SolverException('BRIDGE_UNAVAILABLE', 'bundle evaluation threw: ${result.stringResult}');
    }
    final probe = _runtime.evaluate('typeof PolyPuzzleSolverBridge');
    if (probe.stringResult != 'object') {
      throw SolverException('BRIDGE_UNAVAILABLE', 'PolyPuzzleSolverBridge global not defined after evaluating bundle');
    }
    _bridgeReady = true;
  }

  /// Pre-builds the Solver Engine's wing/flip/case libraries (mirrors the
  /// web app's own warmup() call site) -- call once, well before the first
  /// solve(), per fiveByFiveEdgeSolverEngine.ts's own warmup doc comment.
  void warmup() {
    _requireReady();
    _runtime.evaluate('PolyPuzzleSolverBridge.warmup();');
  }

  /// Runs one solve() call against the frozen Solver Engine.
  ///
  /// Timing note (disclosed, not a defect -- see
  /// docs/IOS_PRODUCT_ARCHITECTURE.md §Bridge Verification, unchanged by
  /// this Flutter pivot): solve() is wall-clock time-budgeted
  /// (PLAN_TIME_BUDGET_MS=1000ms). Its output for a given input can vary
  /// between calls if the host CPU's real execution speed differs --
  /// already established for the web app itself in
  /// docs/SOLVER_LONG_TERM_RELIABILITY_VALIDATION.md's Determinism
  /// Analysis. QuickJS-on-Android's real throughput vs.
  /// JavaScriptCore-on-iOS's was not measured in this Sprint (no device or
  /// simulator available) -- see docs/IOS_DEPLOYMENT_GUIDE.md.
  SolveResult solve(SolverRequest request) {
    _requireReady();
    final requestJson = jsonEncode(request.toJson());
    // jsonEncode(requestJson) below produces a properly escaped JS string
    // literal (JSON string escaping is a strict subset of JS string
    // literal escaping), stored on a throwaway global first rather than
    // interpolated inline, to keep the call expression itself simple and
    // avoid any ambiguity from nested quoting.
    _runtime.evaluate('globalThis.__solverRequestJson = ${jsonEncode(requestJson)};');
    final result = _runtime.evaluate('PolyPuzzleSolverBridge.solve(globalThis.__solverRequestJson);');
    if (result.isError) {
      throw SolverException('BRIDGE_UNAVAILABLE', 'PolyPuzzleSolverBridge.solve() call failed: ${result.stringResult}');
    }

    late final Map<String, dynamic> responseMap;
    try {
      responseMap = jsonDecode(result.stringResult) as Map<String, dynamic>;
    } catch (e) {
      throw SolverException('MALFORMED_RESPONSE', 'response JSON did not decode: $e');
    }

    // The bridge's own contract (PolyPuzzleSolverBridge.ts's solveBridge())
    // is `{ok:true,...} | {ok:false,errorCode,errorMessage}`.
    if (responseMap['ok'] != true) {
      final code = responseMap['errorCode'] as String? ?? 'UNKNOWN';
      final message = responseMap['errorMessage'] as String? ?? '(no message)';
      throw SolverException(code, message);
    }

    try {
      return SolveResult.fromJson(responseMap);
    } catch (e) {
      throw SolverException('MALFORMED_RESPONSE', 'ok:true response did not match SolveResult: $e');
    }
  }

  void _requireReady() {
    if (!_bridgeReady) {
      throw const SolverException('BRIDGE_UNAVAILABLE', 'SolverService.initialize() was not called or failed');
    }
  }

  void dispose() {
    if (_bridgeReady) {
      _runtime.dispose();
    }
  }
}
